### ISSUE TITLE
Celery Beat Redis Broker Leakage

### ISSUE THREAD

**Issue by mikegrima** [celery/celery#4592]
## Checklist

- [X] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [X] I have verified that the issue exists against the `master` branch of Celery.

## Steps to reproduce
Output of `celery -A security_monkey.task_scheduler.tasks.CELERY report`:
```
software -> celery:4.1.0 (latentcall) kombu:4.1.0 py:2.7.12
            billiard:3.5.0.3 redis:2.10.6
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:disabled
... Other things I'm not posting here ...
```

Hello! We are currently using Celery v. 4.1.0 to manage worker jobs in [Security Monkey](https://github.com/Netflix/secuirty_monkey). We initially reported this in #4041, but felt it is worthy of a new issue to track. 

We noticed that the Redis broker has a small memory leak. This leak consumes approximately 400MB a week with our current workload (225 simultaneous worker processes).

Bouncing our scheduler properly clears out the queue as we run `control.purge()`, though, we would prefer to not have to do this periodically.

## Expected behavior
No Redis leakage

## Actual behavior
Redis leakage at about 400MB a week.

![screen shot 2018-03-12 at 11 49 25 am](https://user-images.githubusercontent.com/11035926/37303954-5ec2e7ea-25ed-11e8-82ce-8d4124fe31f4.png)

**Comment by benjolitz** [comment]
I believe I am suffering a variant of this problem. I still run 3.1.25 (I'm trying to migrate away from it, but stability of going to 4 has so far been a surprising experience and I don't have the time).

The symptoms is that the prefork master grows steadily over time in memory usage. When a child fork()s off, it calls `gc.collect` immediately. Decrementing reference counts or any other operation causes the kernel to copy-on-write the memory in question, causing an allocation. When the prefork master (I usually run two per server) exceeds roughly 55-60% memory usage and forks off more than one child (Ah, autoscale you fiend!), the ensuing `gc.collect` causes the server to need to materialize 1.5x the memory it actually has, eats into the swap and eventually gets OOM killed (but only after slowly failing the entire node over the course of an hour).

Using `pyrasite-shell` to break into the prefork master (Note - the prefork master user must be able to read the pyrasite payload file), I noticed the following:

(NOTE: I got from an earlier heap dump session that there was an incredible number of `dict`s, so I knew what to traverse for using objgraph).

```
>>> import collections
>>> count = collections.Counter()
>>> for item in items:
...      for key in item:
...          count[key] += 1
>>> count.most_common(7)
[('tenant_id', 119677), ('headers', 119634), ('delivery_info', 119631), ('reply_to', 119630), ('correlation_id', 119630), (u'utm_campaign', 119624)]
>>> # tenant_id, utm_campaign are mine, part of several tasks that run hourly. 
>>> # prime the trap (remember how many are allocated right now)
>>> objgraph.show_growth(limit=3)
dict    431650   +431650
list    257333   +257333
tuple   131216   +131216
>>> # Invoke one of the tasks (I used a metadata collecting task that runs hourly)
>>> objgraph.show_growth()
tuple               131222        +5
instancemethod      121145        +4
scheduled            59826        +2
weakref               5963        +2
weakcallableproxy    59823        +2
dict                431652        +2
Entry                59823        +2
Request              59814        +1
```

I am still trying to figure out exactly what data structure holds the Requests and/or the `dicts`. For now, I'll restart the prefork master every 24 hours as my charts indicate this is a 1-1.5 day occurrence for "Prefork master grows massively, child forks off of it, child gc's, child gc'ing causes Linux overcommit copy-on-write to instantaneously double referenced memory before freeing, causing excessive swapping for an eventual swap-pocalypse"

**Comment by auvipy** [comment]
please try 4.2rc2 this should be fixed in 4.2 release

**Comment by mikegrima** [comment]
Running a test now. Will report back with my findings.

**Comment by mikegrima** [comment]
@auvipy : Things are looking much better on 4.2rc2. 👍
